### PR TITLE
Generalising the training module to accomodate varying input 

### LINKED ISSUE #19: Generalise the training module
Links to #20 

Convert `train.py` into a flexible script that accommodates different size of inputs (coming from different sources). 

Also make a separate data manipulation file to handle different input sizes.

Also maybe we can add a utility file that contains common functions.

### PR DESCRIPTION
**Pull request by surbhigoel77**
Closes #19 

In order to accomodate data coming from multiple sources, we need to replace the hardcoded inputs with variables whose values are extracted from the dataset during the code run.

Files that need updating:
1. [x] `loaddata.py`
2. [x] `Model.py`
3. [x] `train.py`

In `train.py`, the `.nc` data files are read and the variables are extracted and normalised. This part should be taken out of `train.py` and be kept in a separate `preprocessing.py` file, to keep train.py common for different data sources. `train.py` can be converted into a module instead. We can also have a test notebook in the repo.

### CODE REVIEW

**Comment by surbhigoel77** [comment]
In `Model.py`,  the architecture of the NN is defined. The input/output layer of the NN is hardcoded with the number of input/output variables :
- 93 vertical levels
- 7 input variables that change with vertical levels 
- 4 input variables that do not change with vertical levels
-  2 output variables that change with vertical levels

Assuming that the number of inputs would be different for all three sources, we need to generalise the code for `Model.py` as well (for varying no. of input/output nodes for different sources).

**Comment by surbhigoel77** [comment]
In `train.py`, the `.nc` data files are read and the variables are extracted and normalised. This part should be taken out of `train.py` and be kept in a separate `preprocessing.py` file, to keep train.py common for different data sources.

**Comment by surbhigoel77** [comment]
We can include the normalisation of the variables in the neural network architecture instead of `train.py`

File: newCAM_emulation/Model.py [hunk 1]
```diff
@@ -1,106 +1,103 @@
...
+        -------
+        bool
+            True if training should be stopped, False otherwise.
+        """
+        if validation_loss < self.min_validation_loss:
+            self.min_validation_loss = validation_loss
+            self.counter = 0
+            # if model is not None:
+            #     # torch.save(model.state_dict(), 'conv_torch.pth')
+            #     torch.save(model.state_dict(), 'trained_models/weights_conv')
```
Comment (jatkinson1000) [inline newCAM_emulation/Model.py:136]: Is this dead code, or should it be wrapped in a conditional of some sort, or replaced by a comment?

File: newCAM_emulation/Model.py [hunk 2]
```diff
@@ -1,106 +1,103 @@
...
+        self.hidden_size = hidden_size
+        
+        layers = []
+        input_size = in_ver * ilev + in_nover  
+        for _ in range(hidden_layers):
+            layers.append(nn.Linear(input_size, hidden_size, dtype=torch.float64))
+            layers.append(nn.SiLU())
+            input_size = hidden_size
+        layers.append(nn.Linear(hidden_size, out_ver * ilev, dtype=torch.float64))
+        self.linear_stack = nn.Sequential(*layers)
```
Comment (jatkinson1000) [inline newCAM_emulation/Model.py:68]: A brief comment to summarise what this code is doing for those used to writing it in the previous layout might be useful here.

File: newCAM_emulation/Model.py
Comment (jatkinson1000) [inline newCAM_emulation/Model.py:1]: This file is looking much better, the net is much cleaner and the docstrings really help understand what is going on.

File: newCAM_emulation/Model.py [hunk 3]
```diff
@@ -1,106 +1,103 @@
...
-            loss = loss_fn(pred, Y)
-            avg_loss += loss.item()
-
-    avg_loss /= len(dataloader)
-
-    return avg_loss
-
-
-
+class EarlyStopper:
```
Comment (jatkinson1000) [inline newCAM_emulation/Model.py:86]: It's not immediately clear to me why this should be a class rather than a function, and whether it belons here with the Model, or if it would be better being moved to `train.py` with the training loop/

File: newCAM_emulation/NN_pred.py
Comment (jatkinson1000) [inline newCAM_emulation/NN_pred.py:1]: Am I correct that this has been moved to the `load_data.py` file?

File: newCAM_emulation/loaddata.py [hunk 4]
```diff
@@ -1,45 +1,186 @@
...
+    -------
+    tuple of np.ndarray
+        Input and output arrays for training.
+    """
+    Ncol = normalized_data[variable_names[1]].shape[2]
+    dim_NN = int(in_ver * ilev + in_nover)  
+    dim_NNout = int(out_ver * ilev)
+    x_train = np.zeros([dim_NN, Ncol])
+    y_train = np.zeros([dim_NNout, Ncol])
+    target_var = ['UTGWSPEC', 'VTGWSPEC']
```
Comment (jatkinson1000) [inline newCAM_emulation/loaddata.py:119]: You have made the input names a variable input, but hardcoded the output variables.
Is it worth making both variable?

File: newCAM_emulation/main.py
Comment (jatkinson1000) [inline newCAM_emulation/main.py:1]: I think this presents a much cleaner overview of the development, with details abstracted away into modules.

File: newCAM_emulation/train.py [hunk 5]
```diff
@@ -1,163 +1,97 @@
...
-    split_data = torch.utils.data.random_split(data, [0.75, 0.25], generator=torch.Generator().manual_seed(42))
-    train_dataloader = DataLoader(split_data[0], batch_size=batch_size, shuffle=True)
-    val_dataloader = DataLoader(split_data[1], batch_size=len(split_data[1]), shuffle=True)
-
-     # training
-    early_stopper = EarlyStopper(patience=5, min_delta=0) # Note the hyper parameters.
-    for t in range(epochs):
-        if t % 2 ==0:
-            print(f"Epoch {t+1}\n-------------------------------")
+def train_loop(dataloader, model, loss_fn, optimizer):
```
Comment (jatkinson1000) [inline newCAM_emulation/train.py:4]: Ambiguous name?
This seems to be training a single epoch i.e. a single iteration of the full training loop?

File: newCAM_emulation/train.py [hunk 6]
```diff
@@ -1,163 +1,97 @@
...
-		
+        train_loss = train_loop(train_dataloader, model, criterion, optimizer)
         train_losses.append(train_loss)
-        val_loss = Model.val_loop(val_dataloader, model, nn.MSELoss())
+        val_loss = val_loop(val_dataloader, model, criterion)
         val_losses.append(val_loss)
-        if early_stopper.early_stop(val_loss):
-            print("BREAK!")
+        if early_stopper.early_stop(val_loss, model):
+            # print("BREAK!")
```
Comment (jatkinson1000) [inline newCAM_emulation/train.py:94]: Dead code?
Or, more likely, should we be writing some output to tell the user that training has finished because early stopping criteria was met?

**Review by jatkinson1000 — CHANGES_REQUESTED** [review]
This is looking much better @surbhigoel77 with a few comments on there.
There also appear to be a number of conflicts with the main branch, so perhaps a rebase is needed (though I suggest testing this on a new branch.)

The big thing that is missing is documentation on how to use it which would be nice to allow handing over to collaborators for future re-use. Perhaps #14 can help?

The other thing to consider is some CI though this requires out collaborators understanding how to run formatters and linting. And I don;t think there are any tests yet...?
Was there any way to check that the same results are being generated after the refactor?

File: newCAM_emulation/Model.py (hunk 1 above)
Comment (jatkinson1000) [reply newCAM_emulation/Model.py:136]: Unsure as to it's purpose, ask @yqsun91.

File: newCAM_emulation/Model.py (hunk 3 above)
Comment (jatkinson1000) [reply newCAM_emulation/Model.py:86]: @surbhigoel77 agrees this should be a singe function with inputs, and moved to `train.py`

File: newCAM_emulation/NN_pred.py
Comment (jatkinson1000) [reply newCAM_emulation/NN_pred.py:1]: @surbhigoel77 says mostly to main, but also to other places.

File: newCAM_emulation/loaddata.py [hunk 7]
```diff
@@ -1,45 +1,186 @@
...
+    Prepares the data for training by organizing it into input and output arrays.
+
+    Parameters
+    ----------
+    variable_names : list of str
+        List of variable names.
+    normalized_data : dict
+        Dictionary containing normalized data.
+    ilev : int
+        Number of vertical levels.
```
Comment (jatkinson1000) [inline newCAM_emulation/loaddata.py:107]: Missing some docs for input variables.
